File: frontend/src/api/selection.ts
import { ApiError } from "./editions";

export type SelectionState = "undecided" | "ignored" | "selected";
export type SelectionDecision = "select" | "ignore";

export interface SelectionPublication {
  title: string;
  url: string;
  publisher: string | null;
  role: string;
  published_at: string | null;
}

export interface SelectionProvisionalIoc {
  raw_value: string;
  normalized_value: string | null;
  proposed_type: string;
  declared_type: string | null;
  warnings: string[];
}

export interface SelectionLastDecision {
  id: string;
  decision: SelectionDecision;
  created_at?: string;
  actor_id?: string | null;
}

export interface SelectionItem {
  discovery_subject_id: string;
  title: string;
  summary: string | null;
  presentation: string | null;
  actor_or_campaign: string | null;
  publications: SelectionPublication[];
  candidate_count: number;
  technical_potential: number | null;
  technical_potential_reason: string | null;
  announced_artifacts: string[];
  artifacts?: string[];
  publisher_ioc_count_total: number | null;
  publisher_ioc_counts: number[];
  provisional_ioc_count: number;
  provisional_ioc_type_counts: Record<string, number>;
  provisional_iocs: SelectionProvisionalIoc[];
  uncertainties: string[];
  recommendation: string | null;
  state: SelectionState;
  subject_id: string | null;
  last_decision: SelectionLastDecision | null;
  updated_since_decision: boolean;
  selectable: boolean;
  blocking_reason: string | null;
}

export interface SelectionCounts {
  undecided: number;
  ignored: number;
  selected: number;
  total?: number;
}

export interface SelectionBoard {
  edition_id: string;
  snapshot_id: string | null;
  snapshot_version: number | null;
  counts: SelectionCounts;
  items: SelectionItem[];
  recommendation: string | null;
  fusion_review_count: number;
}

export interface SelectionDecisionInput {
  discovery_subject_id: string;
  decision: SelectionDecision;
  expected_decision_id: string | null;
}

export interface SelectionDecisionRequest {
  snapshot_version: number;
  decisions: SelectionDecisionInput[];
}

/*
 * Wire contract — mirrors `backend/src/cti_app/api/selection.py` field by
 * field. The board above is the UI model: the two schemas differ, so every
 * response goes through `normalizeSelectionBoard` instead of a cast.
 */

interface SelectionPublicationWire {
  url: string;
  title: string;
  publisher: string | null;
  role: string;
  published_at: string | null;
  ioc_declared_count: number | null;
  ioc_visible_count: number | null;
}

interface SelectionProvisionalIocWire {
  raw_value: string;
  normalized_value: string | null;
  declared_type: string | null;
  proposed_type: string;
  warnings: string[];
}

interface SelectionLastDecisionWire {
  id: string;
  action: SelectionDecision;
  snapshot_id: string;
  snapshot_version: number;
  subject_id: string | null;
  actor_id: string;
  occurred_at: string;
}

interface SelectionItemWire {
  discovery_subject_id: string;
  canonical_discovery_subject_id: string;
  title: string;
  summary: string;
  actor_or_campaign: string;
  technical_potential: number;
  technical_potential_reason: string;
  artifacts: string[];
  publications: SelectionPublicationWire[];
  provisional_iocs: SelectionProvisionalIocWire[];
  uncertainties: string[];
  selectable: boolean;
  blocking_reason: string | null;
  effective_state: SelectionState;
  subject_id: string | null;
  recommendation: { recommended: boolean; reason: string | null };
  last_decision: SelectionLastDecisionWire | null;
  updated_since_decision: boolean;
  member_candidate_ids: string[];
}

interface SelectionBoardWire {
  edition_id: string;
  snapshot_id: string | null;
  snapshot_version: number | null;
  items: SelectionItemWire[];
  fusion_review_count: number;
  selected: number;
  ignored: number;
  undecided: number;
}

interface SelectionDecisionWireInput {
  discovery_subject_id: string;
  action: SelectionDecision;
  expected_decision_id: string | null;
}

const RECOMMENDATION_LABELS: Record<string, string> = {
  ioc_signal: "IOC repérés dans les publications.",
};

export function normalizeSelectionBoard(
  board: SelectionBoardWire,
): SelectionBoard {
  return {
    edition_id: board.edition_id,
    snapshot_id: board.snapshot_id,
    snapshot_version: board.snapshot_version,
    counts: {
      undecided: board.undecided,
      ignored: board.ignored,
      selected: board.selected,
      total: board.items.length,
    },
    items: board.items.map(normalizeSelectionItem),
    fusion_review_count: board.fusion_review_count,
    recommendation:
      board.fusion_review_count > 0
        ? `${board.fusion_review_count} fusion(s) à résoudre avant de décider.`
        : null,
  };
}

function normalizeSelectionItem(item: SelectionItemWire): SelectionItem {
  const declaredCounts = item.publications
    .map((publication) => publication.ioc_declared_count)
    .filter((count): count is number => count !== null);
  return {
    discovery_subject_id: item.discovery_subject_id,
    title: item.title,
    summary: item.summary || null,
    presentation: null,
    actor_or_campaign: item.actor_or_campaign || null,
    publications: item.publications.map((publication) => ({
      title: publication.title,
      url: publication.url,
      publisher: publication.publisher,
      role: publication.role,
      published_at: publication.published_at,
    })),
    candidate_count: item.member_candidate_ids.length,
    technical_potential: item.technical_potential,
    technical_potential_reason: item.technical_potential_reason || null,
    announced_artifacts: item.artifacts,
    publisher_ioc_count_total: declaredCounts.length
      ? declaredCounts.reduce((total, count) => total + count, 0)
      : null,
    publisher_ioc_counts: declaredCounts,
    provisional_ioc_count: item.provisional_iocs.length,
    provisional_ioc_type_counts: countIocTypes(item.provisional_iocs),
    provisional_iocs: item.provisional_iocs.map((ioc) => ({
      raw_value: ioc.raw_value,
      normalized_value: ioc.normalized_value,
      proposed_type: ioc.proposed_type,
      declared_type: ioc.declared_type,
      warnings: ioc.warnings,
    })),
    uncertainties: item.uncertainties,
    recommendation: item.recommendation.recommended
      ? (RECOMMENDATION_LABELS[item.recommendation.reason ?? ""] ??
        "Recommandé par la politique de sélection.")
      : null,
    state: item.effective_state,
    subject_id: item.subject_id,
    last_decision: item.last_decision
      ? {
          id: item.last_decision.id,
          decision: item.last_decision.action,
          created_at: item.last_decision.occurred_at,
          actor_id: item.last_decision.actor_id,
        }
      : null,
    updated_since_decision: item.updated_since_decision,
    selectable: item.selectable,
    blocking_reason: item.blocking_reason,
  };
}

function countIocTypes(
  iocs: readonly SelectionProvisionalIocWire[],
): Record<string, number> {
  const counts: Record<string, number> = {};
  for (const ioc of iocs) {
    counts[ioc.proposed_type] = (counts[ioc.proposed_type] ?? 0) + 1;
  }
  return counts;
}

export async function fetchSelectionBoard(
  editionId: string,
): Promise<SelectionBoard> {
  return normalizeSelectionBoard(
    await request<SelectionBoardWire>(selectionUrl(editionId)),
  );
}

export async function confirmSelectionDecisions(
  editionId: string,
  payload: SelectionDecisionRequest,
  idempotencyKey: string,
): Promise<SelectionBoard> {
  // The API names the verb `action`; the board keeps calling it a decision.
  const body: {
    snapshot_version: number;
    decisions: SelectionDecisionWireInput[];
  } = {
    snapshot_version: payload.snapshot_version,
    decisions: payload.decisions.map((decision) => ({
      discovery_subject_id: decision.discovery_subject_id,
      action: decision.decision,
      expected_decision_id: decision.expected_decision_id,
    })),
  };
  return normalizeSelectionBoard(
    await request<SelectionBoardWire>(selectionUrl(editionId, "/decisions"), {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "Idempotency-Key": idempotencyKey,
      },
      body: JSON.stringify(body),
    }),
  );
}

function selectionUrl(editionId: string, suffix = ""): string {
  return `/api/editions/${encodeURIComponent(editionId)}/selection${suffix}`;
}

async function request<T>(url: string, init?: RequestInit): Promise<T> {
  const response = await fetch(url, init);
  if (response.ok) return (await response.json()) as T;

  const body = (await response.json().catch(() => null)) as {
    detail?: { code?: string; message?: string } | string;
  } | null;
  const detail = body?.detail;
  const message =
    typeof detail === "string"
      ? detail
      : detail?.message || "La sélection n’a pas pu être enregistrée.";
  const code =
    typeof detail === "object" && detail?.code
      ? detail.code
      : "selection_error";
  throw new ApiError(message, code, response.status);
}

File: backend/src/cti_app/api/selection.py
from __future__ import annotations

from datetime import datetime
from typing import Any, NoReturn
from uuid import UUID

from fastapi import APIRouter, Header, HTTPException, Request
from pydantic import BaseModel, ConfigDict, Field

from cti_app.application.identity import IdentityProvider
from cti_app.application.selection import (
    SelectionBoard,
    SelectionDecisionCommand,
    SelectionDecisionStaleError,
    SelectionEditionArchivedError,
    SelectionError,
    SelectionIdempotencyConflictError,
    SelectionInvalidCommandError,
    SelectionItem,
    SelectionLastDecision,
    SelectionNotFoundError,
    SelectionRecommendation,
    SelectionService,
    SelectionSnapshotStaleError,
    SelectionSubjectAlreadyMaterializedError,
)
from cti_app.domain.selection import SelectionAction
from cti_app.logging import get_correlation_id

selection_router = APIRouter(prefix="/api/editions/{edition_id}/selection", tags=["selection"])


class SelectionRecommendationView(BaseModel):
    model_config = ConfigDict(extra="forbid")

    recommended: bool
    reason: str | None


class SelectionLastDecisionView(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: UUID
    action: SelectionAction
    snapshot_id: UUID
    snapshot_version: int
    subject_id: UUID | None
    actor_id: str
    occurred_at: datetime


class SelectionItemView(BaseModel):
    model_config = ConfigDict(extra="forbid")

    discovery_subject_id: UUID
    canonical_discovery_subject_id: UUID
    title: str
    summary: str
    actor_or_campaign: str
    technical_potential: int
    technical_potential_reason: str
    artifacts: list[str]
    publications: list[Any]
    provisional_iocs: list[Any]
    uncertainties: list[str]
    selectable: bool
    blocking_reason: str | None
    effective_state: str
    subject_id: UUID | None
    recommendation: SelectionRecommendationView
    last_decision: SelectionLastDecisionView | None
    updated_since_decision: bool
    member_candidate_ids: list[UUID]


class SelectionBoardView(BaseModel):
    model_config = ConfigDict(extra="forbid")

    edition_id: UUID
    snapshot_id: UUID | None
    snapshot_version: int | None
    items: list[SelectionItemView]
    fusion_review_count: int
    selected: int
    ignored: int
    undecided: int


class SelectionDecisionRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    discovery_subject_id: UUID
    action: SelectionAction
    expected_decision_id: UUID | None = None


class SelectionDecisionsRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    snapshot_version: int = Field(ge=1)
    decisions: list[SelectionDecisionRequest] = Field(min_length=1)


@selection_router.get("", response_model=SelectionBoardView)
async def read_selection_board(edition_id: UUID, request: Request) -> SelectionBoardView:
    service: SelectionService = request.app.state.selection_service
    try:
        board = await service.board(edition_id)
    except Exception as exc:
        _raise_selection_error(exc)
    return _board_view(board)


@selection_router.post("/decisions", response_model=SelectionBoardView)
async def decide_selection(
    edition_id: UUID,
    payload: SelectionDecisionsRequest,
    request: Request,
    idempotency_key: str | None = Header(default=None, alias="Idempotency-Key"),
) -> SelectionBoardView:
    if idempotency_key is None or not idempotency_key.strip():
        raise HTTPException(
            status_code=422,
            detail={
                "code": "selection_idempotency_key_required",
                "message": "Idempotency-Key is required and must not be empty.",
            },
        )

    provider: IdentityProvider = request.app.state.identity_provider
    try:
        identity = await provider.current()
    except Exception as exc:
        raise HTTPException(
            status_code=404,
            detail={"code": "identity_not_found", "message": "Identity not found."},
        ) from exc

    correlation_id = get_correlation_id()
    commands = tuple(
        SelectionDecisionCommand(
            discovery_subject_id=decision.discovery_subject_id,
            action=decision.action,
            expected_decision_id=decision.expected_decision_id,
            actor_id=identity.actor_id,
            correlation_id=correlation_id,
            idempotency_key=idempotency_key.strip(),
        )
        for decision in payload.decisions
    )
    service: SelectionService = request.app.state.selection_service
    try:
        board = await service.decide_many(
            edition_id,
            commands,
            snapshot_version=payload.snapshot_version,
        )
    except Exception as exc:
        _raise_selection_error(exc)
    return _board_view(board)


def _raise_selection_error(exc: Exception) -> NoReturn:
    if isinstance(
        exc,
        (
            SelectionSnapshotStaleError,
            SelectionDecisionStaleError,
            SelectionSubjectAlreadyMaterializedError,
            SelectionEditionArchivedError,
            SelectionIdempotencyConflictError,
        ),
    ):
        raise HTTPException(
            status_code=409,
            detail={"code": exc.code, "message": str(exc)},
        ) from exc
    if isinstance(exc, SelectionNotFoundError):
        raise HTTPException(
            status_code=404,
            detail={"code": "selection_not_found", "message": str(exc)},
        ) from exc
    if isinstance(exc, SelectionInvalidCommandError):
        raise HTTPException(
            status_code=422,
            detail={"code": exc.code, "message": str(exc)},
        ) from exc
    if isinstance(exc, SelectionError):
        raise HTTPException(
            status_code=422,
            detail={"code": exc.code, "message": str(exc)},
        ) from exc
    raise HTTPException(
        status_code=500,
        detail={"code": "selection_error", "message": "Selection request failed."},
    ) from exc


def _board_view(board: SelectionBoard) -> SelectionBoardView:
    return SelectionBoardView(
        edition_id=board.edition_id,
        snapshot_id=board.snapshot_id,
        snapshot_version=board.snapshot_version,
        items=[_item_view(item) for item in board.items],
        fusion_review_count=board.fusion_review_count,
        selected=board.selected,
        ignored=board.ignored,
        undecided=board.undecided,
    )


def _item_view(item: SelectionItem) -> SelectionItemView:
    return SelectionItemView(
        discovery_subject_id=item.discovery_subject_id,
        canonical_discovery_subject_id=item.canonical_discovery_subject_id,
        title=item.title,
        summary=item.summary,
        actor_or_campaign=item.actor_or_campaign,
        technical_potential=item.technical_potential,
        technical_potential_reason=item.technical_potential_reason,
        artifacts=list(item.artifacts),
        publications=list(item.publications),
        provisional_iocs=list(item.provisional_iocs),
        uncertainties=list(item.uncertainties),
        selectable=item.selectable,
        blocking_reason=item.blocking_reason,
        effective_state=item.effective_state,
        subject_id=item.subject_id,
        recommendation=_recommendation_view(item.recommendation),
        last_decision=_last_decision_view(item.last_decision),
        updated_since_decision=item.updated_since_decision,
        member_candidate_ids=list(item.member_candidate_ids),
    )


def _recommendation_view(recommendation: SelectionRecommendation) -> SelectionRecommendationView:
    return SelectionRecommendationView(
        recommended=recommendation.recommended,
        reason=recommendation.reason,
    )


def _last_decision_view(
    decision: SelectionLastDecision | None,
) -> SelectionLastDecisionView | None:
    if decision is None:
        return None
    return SelectionLastDecisionView(
        id=decision.id,
        action=decision.action,
        snapshot_id=decision.snapshot_id,
        snapshot_version=decision.snapshot_version,
        subject_id=decision.subject_id,
        actor_id=decision.actor_id,
        occurred_at=decision.occurred_at,
    )
